Improve contributing doc

dai-chen opened the pull request:
*Issue #, if available:*

*Description of changes:* Improve contributing doc by adding a new developing doc with full instructions for developer. Please have a look at https://github.com/dai-chen/sql/blob/new-contributing-doc/docs/developing.rst


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

aetter commented on docs/developing.rst:
```diff
@@ -0,0 +1,350 @@
+
+JDK 12 is required to build the plugin because of the dependency on Elasticsearch test framework in our integration test. So you must have a JDK 12 installation on your machine. After the installation, please configure JAVA_HOME environment variable accordingly. If everything goes right, you should be able to see sample output as below on MacOS::
+
+ $ echo $JAVA_HOME
```
So I know RST supports any level of indentation, but a single space makes it pretty hard to differentiate the code block from the surrounding text. I recommend three spaces so that you're consistent with the level of spacing that numbered lists require.

```rst
So here's a paragraph.

#. Then we have an autonumbered list.

   Here's a paragraph within a list.
   Note that you need three spaces here.

#. Another entry.

Now let's add a code block::

   Note how this block lines up with my lists.
   I get that three spaces are awkward, so do this however you'd like.

Here's another paragraph.

::

   Here's a code block that uses the alternate style.
   Again, I think three spaces makes the most sense given this syntax.

And a final paragraph.

- Yes, I know that unordered lists.

  Only need two spaces for follow-on paragraphs.

- Oh well.
```

dai-chen replied:
Good idea. Will do for all the code blocks.

aetter commented on docs/developing.rst:
```diff
@@ -0,0 +1,350 @@
+Getting Source Code
+===================
+
+Now you can check out the code from your forked GitHub repository and create a new branch for your bug fix or enhancement work.::
```
Remove period since the sentence gets punctuated with a colon.

aetter commented on docs/developing.rst:
```diff
@@ -0,0 +1,350 @@
+ $ git clone https://github.com/<your_account>/sql.git
+ $ git checkout -b <branch_name>
+
+If there is update in master or you want to keep the forked repository long living, you can sync it by following the instructions: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork. Basically you just need to pull latest changes from upstream master once you add it for the first time.::
```
Remove final period before ::.

aetter commented on docs/developing.rst:
```diff
@@ -0,0 +1,350 @@
+Configuring IDEs
+================
+
+You can develop the plugin in your favorite IDEs such as Eclipse and IntelliJ IDEs. Before start making any code change, you may want to configure your IDEs. In this section we take Intellij IDEA for example to show you how to get it ready for development soon.
```
In this section, we show how to get IntelliJ up and running.

aetter commented on docs/developing.rst:
```diff
@@ -0,0 +1,350 @@
+Remote Debugging
+----------------
+
+Firstly you need to add the following configuration to the JVM used by your IDE. For Intellij IDEA, it should be added to <ES installation>/config/jvm.options file. After configuring this, an agent in JVM will listen on the port when your Elasticsearch bootstraps and wait for IDE debugger to connect. So you should be able to debug by setting up a “Remote Run/Debug Configuration”::
```
```rst
``<es-installation>/config/jvm.options``
```

aetter commented on docs/developing.rst:
```diff
@@ -0,0 +1,350 @@
+  * permissions and limitations under the License.
+  */
+
+For example, here is the instructions for Intellij IDEA https://www.jetbrains.com/help/idea/copyright.html to add a Template and Scope profile using the template.
```
```rst
For example, `here are the instructions for adding copyright profiles in IntelliJ IDEA <https://www.jetbrains.com/help/idea/copyright.html>`__.
```

aetter commented on docs/developing.rst:
```diff
@@ -0,0 +1,350 @@
+| ./gradlew integTestRunner       |  Run all integration test (this takes time).              |
++---------------------------------+-----------------------------------------------------------+
+
+For `test` and `integTestRunner`, you can use —tests “UT full path” to run a task individually. For example ./gradlew test --tests “com.amazon.opendistroforelasticsearch.sql.unittest.LocalClusterStateTest”.
```
Double `` for RST inline code.

aetter commented on docs/developing.rst:
```diff
@@ -0,0 +1,350 @@
+
+Most of the time you just need to run ./gradlew build which will make sure you pass all checks and testing. While you’re developing, you may want to run specific Gradle task only. In this case, you can run ./gradlew with task name which only triggers the task along with those it depends on. Here is a list for common tasks:
+
++---------------------------------+-----------------------------------------------------------+
```
So you *can* do this, but yeesh is it hard to maintain. I recommend list tables: http://docutils.sourceforge.net/0.4/docs/ref/rst/directives.html#list-table

aetter commented on docs/developing.rst:
```diff
@@ -0,0 +1,350 @@
+3. Restart Elasticsearch cluster to take it effect.
+
+
+To automate this common task, you can prepare an all-in-one command for reuse. Below is a sample command for MacOS::
```
macOS

aetter commented on docs/developing.rst:
```diff
@@ -0,0 +1,350 @@
+                 post("SELECT ...")
+             )
+         );
+     }
```
Not sure if you meant to close out the class here, but we're missing a }.

dai-chen replied:
Good catch. I missed it. :)

dai-chen commented:
> Hi @dai-chen I added a bunch of total freaking nitpicks. Feel free to ignore all of them, because I think the document looks terrific.

Your comments were really helpful and you responded so quick! I've fixed all. Thanks a lot!

dai-chen commented:
> thanks @dai-chen for the quick and awesome turnaround on the doc.
> please also add a link out to this doc from the README documentation section under, https://github.com/opendistro-for-elasticsearch/sql#documentation
> 
> something in the following format:
> Looking to contribute ? Read the instructions on [Development Guide](link to this doc) and then submit a patch!

Sure, adding it now!

> thanks @dai-chen for the quick and awesome turnaround on the doc.
> please also add a link out to this doc from the README documentation section under, https://github.com/opendistro-for-elasticsearch/sql#documentation
> 
> something in the following format:
> Looking to contribute ? Read the instructions on [Development Guide](link to this doc) and then submit a patch!

Sure, will add it now. Thanks!

chloe-zh commented on docs/developing.rst:
```diff
@@ -0,0 +1,354 @@
+   #Clean and rebuild
+   $ ./gradlew clean
+   $ ./gradlew build
+
```
What about adding the hint of `./gradlew assemble` as well?

chloe-zh replied:
Never mind, I saw it in the following content. :)
